stdlib/digest.ml

```diff
@@ -55,68 +55,94 @@ module type S = sig
   val input : in_channel -> t
   val to_hex : t -> string
   val from_hex : string -> t
+  type state
+  val create : unit -> state
+  val get_digest : state -> t
+  val add_string : state -> string -> unit
+  val add_bytes : state -> bytes -> unit
+  val add_substring : state -> string -> int -> int -> unit
+  val add_subbytes : state -> bytes -> int -> int -> unit
+  val add_channel : state -> in_channel -> int -> unit
 end
 
-(* BLAKE2 hashing, parameterized by hash size *)
+(* Interface for a low-level hash function *)
 
-module BLAKE2 (X: sig val hash_size : int end) : S = struct
+module type HASH = sig
+  type state
+  val hash_size : int
+  val create : unit -> state
+  val update : state -> string -> int -> int -> unit
+  val final : state -> string
+  val unsafe_string : string -> int -> int -> string
```

Comment:
The performance benefit of `unsafe_string` is that no context is allocated as a custom block on the OCaml heap, and then painstakingly collected by the GC; the short-lived context is stack-allocated.  That's what the old MD5 implementation did. 

`unsafe_string` should have been `unsafe_substring`, semantically.  And of course `update` is unsafe (= doesn't check bounds).  Since this is an internal API and bounds checking is performed by the callers, this is fine.  So I settled on `update` and `substring` as the two function names.